integration-tests/src/bin/start_mock_network.rs

```diff
@@ -0,0 +1,89 @@
+use near_actix_test_utils::run_actix;
+use near_chain_configs::GenesisValidationMode;
+use near_client::GetBlock;
+use near_crypto::{InMemorySigner, KeyType};
+use near_logger_utils::init_integration_logger;
+use near_network::test_utils::WaitOrTimeoutActor;
+use near_primitives::types::BlockHeight;
+use std::time::Duration;
+
+/// Program to start a testing environment for one client and a mock network environment
+/// The mock network simulates the entire network by reading a pre-generated chain history
+/// on storage and responds to the client's network requests.
+#[derive(Clap)]
+struct Cli {
+    /// Existing home dir for the pre-generated chain history. For example, you can use
+    /// the home dir of a near node.
+    chain_history_home_dir: String,
+    /// Home dir for the new client that will be started. If not specified, the binary will
+    /// generate a temporary directory
+    client_home_dir: Option<String>,
+    /// Simulated network delay (in ms)
+    #[clap(short = 'd', long, default_value = "100")]
+    network_delay: u64,
+    /// If true, the mock network simulates the client in syncing mode, otherwise,
+    /// the client and its simulated peers will all start from the genesis block
+    #[clap(short = 'S', long)]
+    sync: bool,
+    /// Target height that the client should sync to before stopping. If not specified,
+    /// use the height of the last block in chain history
+    #[clap(short = 'h', long)]
+    target_height: Option<BlockHeight>,
+}
+
+fn main() {
+    init_integration_logger();
+    let args = Cli::parse();
+    let home_dir = Path::new(&args.chain_history_home_dir);
+    let mut near_config = nearcore::config::load_config(home_dir, GenesisValidationMode::Full);
+    near_config.validator_signer = None;
+    near_config.client_config.min_num_peers = 1;
+    let signer =
+        InMemorySigner::from_random("mock_network_node".parse().unwrap(), KeyType::ED25519);
+    near_config.network_config.public_key = signer.public_key;
+    near_config.network_config.secret_key = signer.secret_key;
+    near_config.client_config.tracked_shards =
+        (0..near_config.genesis.config.shard_layout.num_shards()).collect();
+
+    let tempdir = tempfile::Builder::new().prefix("mock_network_node").tempdir().unwrap();
+    let client_home_dir =
+        args.client_home_dir.unwrap_or(String::from(tempdir.path().to_str().unwrap()));
+    let sync_mode = if args.sync { SyncMode::Sync } else { SyncMode::NoSync };
+    let network_delay = Duration::from_millis(args.network_delay);
+    run_actix(async move {
+        let (mock_network, _client, view_client) = setup_mock_network(
+            Path::new(&client_home_dir),
+            home_dir,
+            &near_config,
+            sync_mode,
+            network_delay,
+            args.target_height,
+        );
+        let target_height =
+            async { mock_network.send(GetChainTargetBlockHeight).await }.await.unwrap();
```

Comment:
why don't we use args.target_height?

Reply:
We pass `args.target_height` to `setup_mock_network` so the `GetChainTargetBlockHeight` function will return information with target height incorporated. 

The reason is that `args.target_height` could exceed the the pre-generated chain history, and we want to use the minimum among the two values for termination condition.